modules/files/find.py: use search over match

ritzk opened the pull request:
##### SUMMARY
  - allows for a user to check for a match anywhere in the string
  - https://docs.python.org/2/library/re.html#search-vs-match

This allows one to search for a regex over a string, and not for a pattern from the start of a string.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
modules/files/find.py

##### ANSIBLE VERSION
```
ansible 2.6.0 (find-using-search a142f60407) last updated 2018/03/15 12:33:23 (GMT -400)
  config file = /home/rkhadgar/.ansible.cfg
  configured module search path = [u'/home/rkhadgar/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/rkhadgar/Projects/github/ansible/ansible/lib/ansible
  executable location = /home/rkhadgar/Projects/github/ansible/ansible/bin/ansible
  python version = 2.7.14 (default, Feb 27 2018, 20:43:24) [GCC 7.3.1 20180130 (Red Hat 7.3.1-2)]
```


##### ADDITIONAL INFORMATION
find fails to find file containing regular expression, anywhere in the line.


```
# does not return a mach
$ ansible localhost -m find -a"path='/Projects/github/ansible/ansible contains='map For Ansible by'"

# Returns a match, post patch
$ ansible localhost -m find -a"path='/Projects/github/ansible/ansible contains='map For Ansible by'"
localhost | SUCCESS => {
    "changed": false, 
    "examined": 35, 
    "files": [
        {
             ...
            "path": "/Projects/github/ansible/ansible/ROADMAP.rst", 
# file content
$ head ROADMAP.rst 
*******************************
Roadmap For Ansible by Red Hat
*******************************
...

```

samdoran commented:
This is a good change. However, this module is `stableinterface`, meaning we can't make backwards incompatible changes without a deprecation period. For now, make this an option that can be toggled rather than changing the default.

ritzk commented:
@samdoran  Updated, with deprecated warning.

ritzk committed:
modules/files/find.py: allow regex match anywhere in line

  - new option contains_anywhere to switch between search and match
    https://docs.python.org/2/library/re.html#search-vs-match

samdoran commented on lib/ansible/modules/files/find.py:
```diff
@@ -345,6 +365,9 @@ def main():
     params = module.params
 
+    if params['contains'] and not params['contains_anywhere']:
+        module.deprecate("contains will default to using contains_anywhere as yes in future.", version='3.0')
```
The deprecation version will be 2.11, not 3.0

samdoran replied:
Use True instead of yes just to avoid ambiguity. YAML types are rather fuzzy, True and False are less ambiguous.

samdoran commented on lib/ansible/modules/files/find.py:
```diff
@@ -46,7 +46,17 @@
         description:
-            - One or more regex patterns which should be matched against the file content.
+            - match a regex pattern from the beginning of a line, in a file.
+            - if contains_anywhere is set to yes, match a regex pattern against line in a file.
```
Wrap `contains_anywhere` in `C(contains_anywhere)` so it gets code formatted in the HTML docs.

samdoran replied:
Also capitalize the first word in both of these sentences, please.

samdoran commented on lib/ansible/modules/files/find.py:
```diff
@@ -46,7 +46,17 @@
     contains:
         description:
-            - One or more regex patterns which should be matched against the file content.
+            - match a regex pattern from the beginning of a line, in a file.
```
Comma isn't needed --> 'beginning of a line in a file'

samdoran commented on lib/ansible/modules/files/find.py:
```diff
@@ -46,7 +46,17 @@
+        description:
+            - If no, match the contains regex pattern from the start of a line.
+            - If yes, match the contains regex pattern against a line.
+                - Please use '^' in 'contains' to search for match from the start of a line
```
Wrap `no`, `yes`, `^`, and `contains` in `C()` so they are code formatted in the HTML docs.

bcoca replied:
this should not be a new option, the initial approach to change to 'search' was correct, this might be 'stableinterface' but the current behaviour is a bug.

samdoran commented on lib/ansible/modules/files/find.py:
```diff
@@ -46,7 +46,17 @@
+            - If no, match the contains regex pattern from the start of a line.
+            - If yes, match the contains regex pattern against a line.
+                - Please use '^' in 'contains' to search for match from the start of a line
+                - This option will  default to yes, in future.
```
"...in version 2.11"

samdoran commented on lib/ansible/modules/files/find.py:
```diff
@@ -329,6 +348,7 @@ def main():
             patterns=dict(type='list', default=['*'], aliases=['pattern']),
             excludes=dict(type='list', aliases=['exclude']),
             contains=dict(type='str'),
+            contains_anywhere=dict(type='bool', default='no'),
```
Use a boolean value rather than a string for `default`, i.e., `default=False`.

jborean93 commented:
This has been implemented by https://github.com/ansible/ansible/pull/71083 using the `read_whole_file` option.
